src/ledger/LedgerState.cpp

```diff
@@ -0,0 +1,801 @@
+
+    LedgerEntryChanges changes;
+    for (auto const& state : mState)
+    {
+        // TODO(jonjove): Should output meta if *entry == *previous?
+        auto const& entry = state.second->ignoreInvalid().entry();
+        auto const& previous = state.second->ignoreInvalid().previousEntry();
+        if (previous)
+        {
+            changes.emplace_back(LEDGER_ENTRY_STATE);
+            changes.back().state() = *previous;
+            if (entry)
+            {
+                changes.emplace_back(LEDGER_ENTRY_UPDATED);
+                changes.back().updated() = *entry;
+            }
+            else
+            {
+                changes.emplace_back(LEDGER_ENTRY_REMOVED);
+                changes.back().removed() = state.first;
+            }
+        }
+        else
+        {
+            if (entry)
+            {
+                changes.emplace_back(LEDGER_ENTRY_CREATED);
+                changes.back().created() = *entry;
+            }
+        }
+    }
+    return changes;
+}
+
+std::vector<LedgerEntry>
+LedgerState::getLiveEntries()
+{
+    assert(mRoot);
+    assert(!mChild);
+
+    std::vector<LedgerEntry> entries;
+    for (auto const& state : mState)
+    {
+        auto le = state.second->ignoreInvalid().entry();
+        if (le)
+        {
+            entries.push_back(*le);
+        }
+    }
+    return entries;
+}
+
+std::vector<LedgerKey>
+LedgerState::getDeadEntries()
+{
+    assert(mRoot);
+    assert(!mChild);
+
+    std::vector<LedgerKey> keys;
+    for (auto const& state : mState)
+    {
+        if (state.second->ignoreInvalid().entry() &&
+            !state.second->ignoreInvalid().previousEntry())
+        {
+            keys.push_back(state.first);
+        }
+    }
+    return keys;
+}
+
+void
+LedgerState::commit(std::function<void()> f)
+{
+    assert(mParent || mRoot);
+    assert(!mChild);
+
+    if (mParent)
+    {
+        assert(!f);
+        mergeStateIntoParent();
+        mergeHeaderIntoParent();
+        mParent->mChild = nullptr;
+        mParent = nullptr;
+    }
+    else
+    {
+        soci::transaction sqlTx(mRoot->getDatabase().getSession());
+        try
+        {
+            mergeStateIntoRoot();
+            mergeHeaderIntoRoot();
+            if (f)
+            {
+                f();
+            }
+            sqlTx.commit();
+        }
+        catch (...)
+        {
+            mRoot->getCache().clear();
+            throw;
+        }
+
+        mRoot->hasChild(false);
+        mRoot = nullptr;
+    }
+
+    mHeader.reset();
+    mState.clear();
+    mLoadBestOfferContext.clear();
+}
+
+void
+LedgerState::mergeStateIntoParent()
+{
+    for (auto const& state : mState)
+    {
+        state.second->invalidate();
+        auto const& entry = state.second->ignoreInvalid().entry();
+
+        auto iter = mParent->mState.find(state.first);
+        if (iter == mParent->mState.end())
+        {
+            StateEntry ler(new LedgerEntryReference(entry, nullptr));
```

Comment:
since C++ I don't like seeing naked new in code, what about make_shared?

Reply:
`std::make_shared` can only be used when objects have public constructors (since otherwise `std::make_shared` does not have access to the constructor). But `LedgerEntryReference` and `LedgerHeaderReference` have private constructors as they can only be constructed by `LedgerState`. If you prefer it, we can promote this type of code into a method of its own. For example, something like `StateEntry makeSharedLedgerEntryReference(...)`. I think this is potentially a good idea, since it will mean that naked `new` will be in one place where we can have a clarifying comment explaining the necessity.